feat: Add mock servers for DGX H100, H200, and B200 systems

fabiendupont opened the pull request:
This commit adds comprehensive mock implementations for three additional NVIDIA GPU architectures to support multi-architecture testing:

- dgxh100: 8-GPU DGX H100 system with Hopper architecture
- dgxh200: 8-GPU DGX H200 system with Hopper architecture
- dgxb200: 8-GPU DGX B200 system with Blackwell architecture

Each mock server provides:
- Realistic device properties (memory, compute capability, architecture)
- Complete MIG profile support with architecture-specific configurations
- Full NVML interface implementation following existing dgxa100 patterns
- Thread-safe operations with proper locking
- Comprehensive MIG placement and profile information

Key Features:
- H100/H200: 141GB memory, Hopper architecture, enhanced MIG profiles
- B200: 192GB memory, Blackwell architecture, current-gen capabilities
- Consistent API with existing dgxa100 mock server
- Proper error handling and return codes
- Realistic PCI bus IDs and device UUIDs

This enables comprehensive testing across multiple GPU architectures without requiring physical hardware.

ArangoGutierrez reviewed (commented):
First pass looks good, I'll run some manual tests and provide feedback

elezar commented on pkg/nvml/mock/dgxb200/dgxb200.go:
```diff
@@ -0,0 +1,473 @@
+type Server struct {
+	mock.Interface
+	mock.ExtendedInterface
+	Devices           [8]nvml.Device
```
Do the GB200 nodes not only have 4 devices? Or are these each addressable as 2?

elezar replied:
I think the difference here is that this is a B200 and not a GB200 -- the latter has 4.

ArangoGutierrez replied:
Ah yeah, but here (this PR) is a B200

elezar commented on pkg/nvml/mock/dgxb200/dgxb200.go:
```diff
@@ -0,0 +1,473 @@
+	CudaDriverVersion int
+}
+
+type Device struct {
```
Question: Why do we need a new type. Are the properties not expected to be the same across different devices?

ArangoGutierrez replied:
I think @fabiendupont intend was to create a folder for each GPU type (given that current A100 implementation is a folder named `a100`) If we were to add more GPU types, then we should rename that folder to a generic name, and that way we don't need to duplicate this much code, simply a file per GPU type and MIG profile. reusing a core set of defined variables

ArangoGutierrez commented on pkg/nvml/mock/dgxb200/dgxb200.go:
```diff
@@ -0,0 +1,473 @@
+	return device
+}
+
+func (d *Device) GetUUID() (string, nvml.Return) {
```
Agree @fabiendupont let's do as https://github.com/NVIDIA/go-nvml/blob/main/pkg/nvml/mock/dgxa100/dgxa100.go#L195  and https://github.com/NVIDIA/go-nvml/blob/main/pkg/nvml/mock/dgxa100/dgxa100.go#L136 for the 3 B200,H100 and H200

ArangoGutierrez reviewed (dismissed):
PR looks now now
@elezar  - @klueska  Ready for your approval

elezar commented on pkg/nvml/mock/README.md:
````diff
@@ -0,0 +1,270 @@
+
+Add new configurations to the appropriate file in `shared/gpus/`:
+```go
+var A100_PCIE_24GB = shared.Config{
````
nit: Do we want a more go-like name?

elezar commented on pkg/nvml/mock/dgxa100/dgxa100.go:
```diff
@@ -17,365 +17,63 @@
-		return nvml.SUCCESS
-	}
+// NewDeviceWithGPU creates a new device with a specific A100 GPU variant
+func NewDeviceWithGPU(gpuConfig shared.Config, index int) *Device {
```
Should this really be defined for this package?

elezar commented on pkg/nvml/mock/shared/shared.go:
```diff
@@ -0,0 +1,426 @@
+		PciDeviceId:        config.PciDeviceId,
+		MIGProfiles:        config.MIGProfiles,
+	}
+	device.SetMockFuncs()
```
Note: Since a `device` is only functional after setting these mocks, should we be exposing it as a public concrete type?

elezar commented on pkg/nvml/mock/dgxh100/dgxh100.go:
```diff
@@ -0,0 +1,79 @@
+func New() *Server {
+	return shared.NewServerFromConfig(shared.ServerConfig{
+		Config:            gpus.H100_SXM5_80GB,
+		GPUCount:          8,
```
Question: Instead of specifying the count, why not pass in a list of devices? This will also make it simpler to handle heterogenous configs.

elezar commented on pkg/nvml/mock/dgxa100/dgxa100.go:
```diff
@@ -17,365 +17,63 @@
-		return gis, nvml.SUCCESS
-	}
+// NewServerWithGPU creates a new server with a specific A100 GPU variant
+func NewServerWithGPU(gpuConfig shared.Config) *Server {
```
Also, what about implementing this as:

```suggestion
func NewServerWithGPUs(gpuConfigs ...shared.Config) *Server {
```
and updagin the `NewServiceFromConfig` implementation to handle this?

elezar reviewed (commented):
Some quick thoughts on this. I will make a more critical pass later this week.

fabiendupont commented on pkg/nvml/mock/README.md:
````diff
@@ -0,0 +1,270 @@
+
+```
+pkg/nvml/mock/
+├── shared/
````
Moved it to internal. Users still have an external API through package-specific wrappers.

elezar reviewed (commented):
Sorry for the delay here. I am revisiting this due to my wanting to add a specific mock for something that I'm working on.

Looking at the orginisation, I think the SPECIFIC GPU implementations are useful -- as is the scaffolding to create a server with a specific list of GPUs or a COUNG of a single GPU type.

In my case, I want to mock a Tegra-based system which includes an iGPU and a dGPU. Here I would imaging instantiating this as:
```
NewServer(WithGPUs(gpus.IGX_THOR, gpus.RTX_PRO_6000))
```
Let's assume that I don't care about the driver version and / or the CUDA version in this case.

WIth this in mind, I have some proposed changes under https://github.com/elezar/go-nvml/tree/feat/add-dgxh100-dgxh200-dgxb200-mocks

I also have a branch: https://github.com/elezar/go-nvml/tree/add-tegra-mocks that adds a tegra (Thor) device.

tariq1890 commented:
@fabiendupont Can you rebase this PR? Also, are you to incorporate the changes suggested by @elezar in his [branch](https://github.com/NVIDIA/go-nvml/compare/main...elezar:go-nvml:add-tegra-mocks)?

fabiendupont commented:
@tariq1890, done. I hope it's still looking good :)

tariq1890 commented on pkg/nvml/mock/README.md:
```diff
@@ -0,0 +1,396 @@
+│       ├── a30.go               # A30 GPU variants (Ampere)
+│       ├── h100.go              # H100 GPU variants (Hopper)
+│       ├── h200.go              # H200 GPU variants (Hopper)
+│       └── b200.go              # B200 GPU variants (Blackwell)
```
Not applicable as these mocks are intended to be exportable.

tariq1890 commented on pkg/nvml/mock/dgxa100/dgxa100.go:
```diff
@@ -17,365 +17,70 @@
+	return NewWithGPUs(gpus.Multiple(8, gpus.A100_SXM4_40GB)...)
 }
 
-func NewDevice(index int) *Device {
```
Btw we depend on this method in `mig-parted` unit test cases. I imported your fork and I can confirm that there are compile errors. Can we retain this method?

tariq1890 replied:
See [here](https://github.com/NVIDIA/mig-parted/blob/f821741da02d6cce626bbad91841c1ad62d7f69c/pkg/mig/mode/nvml_test.go#L43)

tariq1890 commented:
@fabiendupont Multiple commits are okay, but let's ensure that each commit is complete i.e. Each commit brings the repo to a working and usable state. If you'd like squash all of the commits into one commit, that's also fine. Completeness of the commit is key either way.

fabiendupont committed:
feat: Add mock servers for DGX H100, H200, and B200 systems

Add comprehensive mock implementations for three additional NVIDIA GPU
architectures to support multi-architecture testing:

- Introduce gpus/ package with GPU configuration structs (Config,
  MIGProfileConfig) and per-GPU-model definitions (A100, H100, H200,
  B200, A30) including full MIG profile data
- Introduce server/ package with shared Server, Device, GpuInstance,
  and ComputeInstance types using functional options pattern
  (WithGPUs, WithDriverVersion, etc.)
- Refactor dgxa100/ to thin wrapper delegating to server.New(),
  maintaining backward-compatible type aliases and MIG globals
- Add dgxh100/: 8-GPU DGX H100 (Hopper, 80GB HBM3, CUDA 9.0)
- Add dgxh200/: 8-GPU DGX H200 (Hopper, 141GB HBM3e, CUDA 9.0)
- Add dgxb200/: 8-GPU DGX B200 (Blackwell, 180GB HBM3e, CUDA 10.0)
- Sync A100 compute instance profiles with upstream main (REV1
  entries, SharedCopyEngineCount fields)
- Add comprehensive tests for all four DGX packages

Architecture based on design by Evan Lezar.

Co-authored-by: Evan Lezar <elezar@nvidia.com>
Signed-off-by: Fabien Dupont <fdupont@redhat.com>

rh-pre-commit.version: 2.4.0
rh-pre-commit.check-secrets: ENABLED

fabiendupont commented:
@tariq1890 Done. Here's what changed:

- Rebased onto current `main`
- Incorporated @elezar's architectural changes from his [branch](https://github.com/NVIDIA/go-nvml/compare/main...elezar:go-nvml:add-tegra-mocks): `internal/shared/` moved to public `gpus/` + `server/` packages with functional options pattern
- Squashed into a single commit for a clean history
- Synced A100 compute instance profiles with upstream (REV1 entries from acfe0bd)
- Restored `NewDevice`, `NewGpuInstance`, `NewComputeInstance` on `dgxa100` for backward compatibility with `mig-parted` — and added them to `dgxh100`, `dgxh200`, `dgxb200` for consistency
- @elezar credited as co-author

Tegra support is left for @elezar to add in a follow-up — his architecture makes that straightforward.

tariq1890 commented:
Thank you very much for your contribution @fabiendupont ! Really appreciate your patience with this.
